pkg/controller/nstemplatetier/nstemplatetier_controller.go

```diff
@@ -0,0 +1,238 @@
+	err := r.client.Get(context.TODO(), request.NamespacedName, &instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		logger.Error(err, "unable to get the current NSTemplateTier instance")
+		return reconcile.Result{}, err
+	}
+
+	activeTemplateUpdateRequests, err := r.activeTemplateUpdateRequests(instance)
+	if err != nil {
+		logger.Error(err, "unable to get active TemplateUpdateRequests")
+		return reconcile.Result{}, err
+	}
+
+	if activeTemplateUpdateRequests < maxPoolSize {
+		members := r.retrieveMemberClusters()
+	loop:
+		for _, member := range members {
+			// create a TemplateUpdateRequest if active count < max pol size
+			// find a MasterUserRecord which is not already up-to-date
+			// and for which there is no TemplateUpdateRequest yet
+			// fetch by subsets of "maxPoolSize + 1" size until a MasterUserRecord candidate is found
+			murs := toolchainv1alpha1.MasterUserRecordList{}
+			matchingLabels, err := newLabelSelector(instance, *member)
+			if err != nil {
+				logger.Error(err, "unable to get MasterUserRecords to update")
+				return reconcile.Result{}, err
+			}
+			r.client.List(context.Background(), &murs,
+				client.InNamespace(instance.Namespace),
+				client.Limit(maxPoolSize+1),
+				matchingLabels,
+			)
+			logger.Info("listed MasterUserRecords", "count", len(murs.Items), "cluster", member.Name)
+			for _, mur := range murs.Items {
+				for _, account := range mur.Spec.UserAccounts {
+					accountTmpls := account.Spec.NSTemplateSet
+					logger.Info("checking mur account", "name", mur.Name, "tier", accountTmpls.TierName)
+					if accountTmpls.TierName != instance.Name {
+						// skip if the user account is not based on the current Tier
+						continue
+					}
+					// compare all TemplateRefs in the current MUR.UserAccount vs NSTemplateTier instance
+					if !reflect.DeepEqual(templateRefs(accountTmpls), templateRefs(instance.Spec)) {
+						logger.Info("creating a TemplateUpdateRequest to update the MasterUserRecord", "name", mur.Name, "tier", instance.Name)
+						r.client.Create(context.TODO(), &toolchainv1alpha1.TemplateUpdateRequest{
+							ObjectMeta: metav1.ObjectMeta{
+								Name: mur.Name,
+								Labels: map[string]string{
+									toolchainv1alpha1.NSTemplateTierNameLabelKey: instance.Name,
+								},
+							},
+							Spec: toolchainv1alpha1.TemplateUpdateRequestSpec{
+								TierName:           instance.Name,
+								NSTemplateTierSpec: instance.Spec,
+							},
+						})
+						break loop // create a single TemplateUpdateRequest resource per reconcile loop
+					}
+				}
+			}
+		}
+	}
+	logger.Info("done with creating TemplateUpdateRequest resources after update of NSTemplateTier", "tier", instance.Name)
+	return reconcile.Result{}, nil
+}
+
+func newLabelSelector(instance toolchainv1alpha1.NSTemplateTier, member cluster.FedCluster) (client.MatchingLabelsSelector, error) {
```

Comment:
Can we add a comment explaining what this function does and why?

Reply:
sure! done in 135416c